File: sdk-js/src/providers/anthropic.ts
// Copyright 2026 Harsha Krishne Gowda
// SPDX-License-Identifier: Apache-2.0

/** Anthropic provider integration — wraps the Anthropic client to track usage. */

import type { UsageEvent, TokenBudgetOptions } from '../types.js';
import { resolveConfig } from '../config.js';
import { calculateCost } from '../pricing.js';
import { getCurrentTags } from '../context.js';
import { EventTransport } from '../transport.js';

// Shared transport instance
let _transport: EventTransport | null = null;

/**
 * Wrap an Anthropic client to automatically track token usage.
 *
 * Patches `client.messages.create` in-place and returns the same client.
 *
 * @param client - An Anthropic client instance
 * @param options - TokenBudget configuration options
 * @returns The same client, now patched to track usage
 */
export function wrapAnthropic<T extends Record<string, any>>(
  client: T,
  options: TokenBudgetOptions = {},
): T {
  const config = resolveConfig(options);

  if (!config.enabled) return client;

  if (!_transport) {
    _transport = new EventTransport(config);
  }

  const transport = _transport;
  const messages = client.messages as Record<string, any>;

  if (!messages?.create) {
    throw new Error(
      'TokenBudget: client.messages.create not found. Pass an Anthropic client instance.',
    );
  }

  const originalCreate = messages.create.bind(messages);

  messages.create = async function patchedCreate(
    ...args: any[]
  ): Promise<any> {
    const start = performance.now();
    const response = await originalCreate(...args);
    const elapsedMs = Math.round(performance.now() - start);

    try {
      const model: string = response.model ?? 'unknown';
      const usage = response.usage;
      const inputTokens: number = usage?.input_tokens ?? 0;
      const outputTokens: number = usage?.output_tokens ?? 0;
      const totalTokens = inputTokens + outputTokens;
      const costUsd = calculateCost(model, inputTokens, outputTokens);
      const tags = getCurrentTags();

      const event: UsageEvent = {
        provider: 'anthropic',
        model,
        input_tokens: inputTokens,
        output_tokens: outputTokens,
        total_tokens: totalTokens,
        cost_usd: costUsd,
        latency_ms: elapsedMs,
        tags,
        metadata: {},
        timestamp: Date.now() / 1000,
      };

      transport.send(event);
    } catch {
      // Never let tracking errors affect the user
    }

    return response;
  };

  return client;
}

/** Flush and shut down the Anthropic transport. Exported for testing. */
export async function _shutdownAnthropic(): Promise<void> {
  if (_transport) {
    await _transport.shutdown();
    _transport = null;
  }
}

/** Reset the shared transport. Exported for testing. */
export function _resetAnthropic(): void {
  _transport = null;
}

File: sdk/tokenbudget/providers/anthropic.py
# Copyright 2026 Harsha Krishne Gowda
# SPDX-License-Identifier: Apache-2.0

"""Anthropic provider integration."""
from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any

from tokenbudget.providers.base import BaseProvider
from tokenbudget.pricing import calculate_cost
from tokenbudget.types import UsageEvent

if TYPE_CHECKING:
    from tokenbudget.transport import EventTransport

logger = logging.getLogger(__name__)


class AnthropicProvider(BaseProvider):
    """Provider integration for the Anthropic Python client."""

    def detect(self, client: Any) -> bool:
        """Return True if the client is from the anthropic package."""
        return type(client).__module__.startswith("anthropic")

    def extract_event(self, response: Any, latency_ms: int) -> UsageEvent:
        """Extract a UsageEvent from an Anthropic messages response."""
        from tokenbudget.context import get_current_tags

        model = getattr(response, "model", "unknown")
        usage = getattr(response, "usage", None)

        input_tokens = getattr(usage, "input_tokens", 0) if usage else 0
        output_tokens = getattr(usage, "output_tokens", 0) if usage else 0
        total_tokens = input_tokens + output_tokens

        cost = calculate_cost(model, input_tokens, output_tokens)
        tags = get_current_tags()

        return UsageEvent(
            provider="anthropic",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=total_tokens,
            cost_usd=cost,
            latency_ms=latency_ms,
            tags=tags,
        )

    def patch(self, client: Any, transport: "EventTransport") -> None:
        """Monkey-patch client.messages.create to track usage."""
        original_create = client.messages.create

        def patched_create(*args: Any, **kwargs: Any) -> Any:
            start = time.perf_counter()
            response = original_create(*args, **kwargs)
            elapsed_ms = int((time.perf_counter() - start) * 1000)
            try:
                event = self.extract_event(response, latency_ms=elapsed_ms)
                transport.send(event)
            except Exception as exc:
                logger.warning("TokenBudget: failed to track Anthropic event: %s", exc)
            return response

        client.messages.create = patched_create
